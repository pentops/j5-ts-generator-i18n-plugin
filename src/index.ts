import type { Resource, ResourceLanguage, InitOptions } from 'i18next';
import { match, P } from 'ts-pattern';
import {
  type GeneratedSchema,
  PluginBase,
  type PluginConfig,
  type PluginFileGeneratorConfig,
  type GeneratedClientFunction, createObjectLiteral,
} from '@pentops/jsonapi-jdef-ts-generator';
import { camelCase } from 'change-case';
import set from 'lodash.set';
import ts, { factory } from 'typescript';

export const I18NEXT_IMPORT_PATH = 'i18next';
export const I18NEXT_DEFAULT_EXPORT_NAME = 'i18n';
export const I18NEXT_INIT_FUNCTION_NAME = 'init';
export const I18NEXT_USE_FUNCTION_NAME = 'use';

export type TranslationValue = string | null;

export interface Translation {
  // JSON dot notation
  key: string;
  value: TranslationValue;
}

export type I18nPluginTranslationPathGetter = (schema: GeneratedSchema) => string | undefined;

export type I18nPluginTranslationWriter = (
  schema: GeneratedSchema,
  schemaPath: string,
  existingValues?: Map<string, Translation>,
) => Translation[] | undefined;

export interface I18nPluginFileGeneratorConfig extends PluginFileGeneratorConfig {
  language: string;
  namespaceName?: string;
  translationPathOrGetter?: string | I18nPluginTranslationPathGetter;
  translationWriter?: I18nPluginTranslationWriter;
  unmatchedTranslationFromExistingFileHandler?: 'keep' | 'remove' | ((translation: Translation) => Translation | null);
}

export type I18nPluginFileConfigCreator = (
  generatedSchemas: Map<string, GeneratedSchema>,
  generatedClientFunctions: GeneratedClientFunction[],
) => I18nPluginFileGeneratorConfig[];

export const defaultTranslationPathOrGetter: I18nPluginTranslationPathGetter = (schema: GeneratedSchema): string | undefined =>
  match(schema)
    .with({ rawSchema: { oneOf: P.not(P.nullish) } }, () => `oneOf.${schema.generatedName}`)
    .with({ rawSchema: { enum: P.not(P.nullish) } }, () => `enum.${schema.generatedName}`)
    .otherwise(() => undefined);

export const defaultSchemaTranslationWriter: I18nPluginTranslationWriter = (
  schema: GeneratedSchema,
  schemaPath: string,
  existingValues?: Map<string, Translation>,
): Translation[] | undefined =>
  match(schema)
    .with({ rawSchema: { oneOf: P.not(P.nullish) } }, (s) =>
      Array.from(s.rawSchema.oneOf.properties.values()).map((property) => {
        const path = `${schemaPath}.${property.name}`;

        if (existingValues?.has(path)) {
          return existingValues.get(path)!;
        }

        return {
          key: path,
          value: property.name,
        };
      }),
    )
    .with({ rawSchema: { enum: P.not(P.nullish) } }, (s) =>
      s.rawSchema.enum.options.map((value) => {
        const path = `${schemaPath}.${value.name}`;

        if (existingValues?.has(path)) {
          return existingValues.get(path)!;
        }

        return {
          key: path,
          value: value.name,
        };
      }),
    )
    .otherwise(() => undefined);

export interface I18nIndexMiddlewareConfig {
  importPath: string;
  importSpecifier: string;
  isDefault?: boolean;
}

export interface I18nIndexFileConfig extends PluginFileGeneratorConfig {
  addGeneratedResources?: boolean;
  initOptions?: InitOptions;
  middleware?: I18nIndexMiddlewareConfig[];
}

export interface I18nPluginConfig extends PluginConfig<I18nPluginFileGeneratorConfig> {
  files: I18nPluginFileGeneratorConfig[] | I18nPluginFileConfigCreator;
  indexFile?: I18nIndexFileConfig;
}

export class I18nPlugin extends PluginBase<I18nPluginFileGeneratorConfig, I18nPluginConfig> {
  name = 'I18nPlugin';

  // eslint-disable-next-line @typescript-eslint/no-useless-constructor
  constructor(config: I18nPluginConfig) {
    super(config);
  }

  private static parseExistingValue(value: string | undefined): ResourceLanguage | undefined {
    if (!value) {
      return undefined;
    }

    try {
      return JSON.parse(value);
    } catch (e) {
      throw new Error(`I18nPlugin: failed to parse existing value: ${value}. ${e}`);
    }
  }

  private static gatherTranslations(fileData: ResourceLanguage | undefined) {
    const translations = new Map<string, Translation>();

    if (!fileData) {
      return translations;
    }

    const gather = (data: string | Record<string, any>, path: string = '') => {
      if (typeof data === 'object') {
        for (const [key, value] of Object.entries(data)) {
          gather(value, path ? `${path}.${key}` : key);
        }
      } else {
        translations.set(path, { key: path, value: data });
      }
    };

    gather(fileData);

    return translations;
  }

  private static buildResourcesObjectLiteral(providedResources: Resource, generatedResources: Record<string, Record<string, string>>) {
    const mergedResources: Record<string, Record<string, string | object>> = {};
    const allLanguages = Array.from(new Set(Object.keys(providedResources).concat(Object.keys(generatedResources))));

    for (const language of allLanguages) {
      const provided = providedResources[language] || {};
      const generated = generatedResources[language] || {};
      const allNamespaces = Array.from(new Set(Object.keys(provided).concat(Object.keys(generated))));

      if (!mergedResources[language]) {
        mergedResources[language] = {};
      }

      for (const namespace of allNamespaces) {
        const value = match({ provided: provided[namespace], generated: generated[namespace] })
          .with({ generated: P.not(undefined) }, ({ generated: g }) => factory.createIdentifier(g))
          .with({ provided: P.not(undefined) }, ({ provided: p }) => p)
          .otherwise(() => undefined);

        if (value) {
          mergedResources[language][namespace] = value;
        }
      }
    }

    return createObjectLiteral(mergedResources);
  }

  private static getFileNamespace(file: I18nPluginFileGeneratorConfig) {
    return file.namespaceName || camelCase(file.fileName.replace('.json', ''));
  }

  private generateIndexFile() {
    if (!this.pluginConfig.indexFile) {
      return;
    }

    const { addGeneratedResources, initOptions, middleware, ...defaultFileConfig } = this.pluginConfig.indexFile;

    const indexFile = this.createPluginFile(defaultFileConfig);

    indexFile.addManualImport(I18NEXT_IMPORT_PATH, [], [], I18NEXT_DEFAULT_EXPORT_NAME);

    const resourcesByLanguageAndNamespace: Record<string, Record<string, string>> = {};
    const { resources = {}, ...remainingInitOptions } = initOptions || {};

    if (addGeneratedResources) {
      const filesToAdd = this.files.filter((file) => file.getHasContent());

      if (filesToAdd.length) {
        for (const file of filesToAdd) {
          if (!resourcesByLanguageAndNamespace[file.config.language]) {
            resourcesByLanguageAndNamespace[file.config.language] = {};
          }

          const languageResources = resourcesByLanguageAndNamespace[file.config.language]!;
          const namespaceName = I18nPlugin.getFileNamespace(file.config);
          const resourceSpecifier = camelCase(`${file.config.language}_${namespaceName}_NS`);

          languageResources[namespaceName] = resourceSpecifier;

          indexFile.addImportToOtherGeneratedFile(file, undefined, undefined, resourceSpecifier);
        }
      }
    }

    const resourcesObjectLiteral = I18nPlugin.buildResourcesObjectLiteral(resources, resourcesByLanguageAndNamespace);

    // The index file is slightly different from the other files generated by this plugin, so just casting to any for now
    this.files.push(indexFile as any);

    let callExpression: ts.CallExpression | undefined;

    // Configure middleware
    for (const m of middleware || []) {
      indexFile.addManualImport(m.importPath, m.isDefault ? undefined : [m.importSpecifier], [], m.isDefault ? m.importSpecifier : undefined);

      callExpression = factory.createCallExpression(
        factory.createPropertyAccessExpression(
          callExpression || factory.createIdentifier(I18NEXT_DEFAULT_EXPORT_NAME),
          factory.createIdentifier(I18NEXT_USE_FUNCTION_NAME),
        ),
        undefined,
        [factory.createIdentifier(m.importSpecifier)],
      );
    }

    callExpression = factory.createCallExpression(
      factory.createPropertyAccessExpression(
        callExpression || factory.createIdentifier(I18NEXT_DEFAULT_EXPORT_NAME),
        factory.createIdentifier(I18NEXT_INIT_FUNCTION_NAME),
      ),
      undefined,
      [createObjectLiteral({ ...remainingInitOptions, resources: resourcesObjectLiteral })],
    );

    indexFile.addNodes(callExpression, factory.createIdentifier('\n'));
    indexFile.addManualExport(undefined, { namedExports: [I18NEXT_DEFAULT_EXPORT_NAME], typeOnlyExports: [] })
  }

  public async run() {
    for (const file of this.files) {
      const fileData = I18nPlugin.parseExistingValue(file.existingFileContent);
      const existingTranslationsInFile = I18nPlugin.gatherTranslations(fileData);
      const generatedTranslations: Map<string, Translation> = new Map();

      for (const [, schema] of this.generatedSchemas) {
        if (file.isFileForSchema(schema)) {
          const translationPath =
            typeof file.config.translationPathOrGetter === 'function'
              ? file.config.translationPathOrGetter(schema)
              : file.config.translationPathOrGetter || defaultTranslationPathOrGetter(schema);

          if (translationPath) {
            const translationsForSchema = file.config.translationWriter
              ? file.config.translationWriter(schema, translationPath, existingTranslationsInFile)
              : defaultSchemaTranslationWriter(schema, translationPath, existingTranslationsInFile);

            for (const translation of translationsForSchema || []) {
              generatedTranslations.set(translation.key, translation);
            }
          }
        }
      }

      for (const [key, value] of existingTranslationsInFile) {
        if (!generatedTranslations.has(key)) {
          const unmatchedTranslationHandler = file.config.unmatchedTranslationFromExistingFileHandler;

          if (typeof unmatchedTranslationHandler === 'function') {
            const newValue = unmatchedTranslationHandler(value);

            if (newValue) {
              generatedTranslations.set(key, newValue);
            }
          } else if (unmatchedTranslationHandler !== 'remove') {
            generatedTranslations.set(key, value);
          }
        }
      }

      const fileContent: ResourceLanguage = {};

      for (const translation of generatedTranslations.values()) {
        if (translation.value !== null && translation.value !== undefined) {
          set(fileContent, translation.key, translation.value);
        }
      }

      file.setRawContent(JSON.stringify(fileContent, null, 2));
    }

    if (this.pluginConfig.indexFile) {
      this.generateIndexFile();
    }
  }
}
