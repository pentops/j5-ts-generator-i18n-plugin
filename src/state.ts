import { Translation } from './helpers';
import { match, P } from 'ts-pattern';

export interface WrittenTranslation {
  key: string;
  value: string;
  source?:
    | {
        enum: {
          fullGrpcName: string;
          generatedName: string;
          values: Record<string, string>;
        };
      }
    | {
        oneOf: {
          fullGrpcName: string;
          generatedName: string;
          values: string[];
        };
      }
    | {
        polymorph: {
          fullGrpcName: string;
          generatedName: string;
          members?: string[];
        };
      };
}

export interface I18nPluginState {
  translationsWritten: Record<string, WrittenTranslation>;
}

export function buildState(translationsWritten: Record<string, Translation>): I18nPluginState {
  return {
    translationsWritten: Object.entries(translationsWritten).reduce<Record<string, WrittenTranslation>>((acc, [key, value]) => {
      return match(value.source)
        .with({ generatedValueNames: P.not(P.nullish), rawSchema: { enum: { derivedHelperType: P.nullish } } }, (s) => ({
          ...acc,
          [key]: {
            key,
            value: value.value,
            source: {
              enum: {
                fullGrpcName: s.rawSchema.enum.fullGrpcName,
                generatedName: s.generatedName,
                values: Object.fromEntries(s.generatedValueNames.entries()),
              },
            },
          },
        }))
        .with({ generatedValueNames: P.not(P.nullish), rawSchema: { enum: { derivedHelperType: P.not(P.nullish) } } }, (s) => ({
          ...acc,
          [key]: {
            key,
            value: value.value,
            source: {
              oneOf: {
                fullGrpcName: s.rawSchema.enum.fullGrpcName,
                generatedName: s.generatedName,
                values: Array.from(s.generatedValueNames.keys()),
              },
            },
          },
        }))
        .otherwise(() => ({
          ...acc,
          [key]: {
            key,
            value: value.value,
          },
        }));
    }, {}),
  };
}
