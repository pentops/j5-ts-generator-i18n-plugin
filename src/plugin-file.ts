import { BasePlugin, BasePluginFile } from '@pentops/jsonapi-jdef-ts-generator';
import { I18nPluginFileGeneratorConfig } from './plugin';

export class I18nPluginFile extends BasePluginFile<string, I18nPluginFileGeneratorConfig, BasePlugin<any>> {}
