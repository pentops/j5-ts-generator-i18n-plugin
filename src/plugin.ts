import ts, { factory } from 'typescript';
import type { InitOptions, ResourceLanguage } from 'i18next';
import {
  createObjectLiteral,
  type GeneratedClientFunction,
  type GeneratedSchema,
  type Optional,
  BasePlugin,
  type IPluginConfig,
  type IPluginFileConfig,
  type IPluginRunOutput,
  defaultGeneratorFileReader,
  type IWritableFile,
} from '@pentops/jsonapi-jdef-ts-generator';
import { camelCase } from 'change-case';
import setWith from 'lodash.setwith';
import { sortByKey } from '@pentops/sort-helpers';
import {
  buildProspectiveTranslations,
  buildResourcesObjectLiteral,
  defaultConflictHandler,
  defaultDefinedAnySchemaTranslationPathGetter,
  defaultI18nPluginDefinedAnySchemaTranslationWriter,
  defaultNamespaceWriter,
  defaultSchemaTranslationWriter,
  defaultTranslationPathOrGetter,
  gatherTranslations,
  I18N_INIT_OPTS_VAR_NAME,
  I18N_NAMESPACES_TYPE_NAME,
  I18NEXT_DEFAULT_EXPORT_NAME,
  I18NEXT_IMPORT_PATH,
  I18NEXT_INIT_FUNCTION_NAME,
  I18NEXT_USE_FUNCTION_NAME,
  type I18nPluginConflictHandler,
  I18nPluginDefinedAnySchemaTranslationPathGetter,
  I18nPluginDefinedAnySchemaTranslationWriter,
  type I18nPluginTranslationPathGetter,
  type I18nPluginTranslationWriter,
  type NamespaceWriter,
  parseExistingValue,
  type Translation,
} from './helpers';
import { I18nPluginFile } from './plugin-file';
import { buildState, I18nPluginState } from './state';

export interface I18nPluginFileGeneratorConfig<TFileContentType = string> extends Omit<IPluginFileConfig<TFileContentType>, 'exportFromIndexFile'> {
  language: string;
  namespaceName?: string;
  translationPathOrGetter?: I18nPluginTranslationPathGetter;
  translationWriter?: I18nPluginTranslationWriter;
  unmatchedTranslationFromExistingFileHandler?: 'keep' | 'remove' | ((translation: Translation) => Translation | null);
}

export type I18nPluginFileConfigCreator<TFileContentType = string> = (
  generatedSchemas: Map<string, GeneratedSchema>,
  generatedClientFunctions: GeneratedClientFunction[],
) => I18nPluginFileGeneratorConfig<TFileContentType>[];

export interface I18nIndexMiddlewareConfig {
  importPath: string;
  importSpecifier: string;
  isDefault?: boolean;
}

export interface I18nIndexFileConfig<TFileContentType = string> extends IPluginFileConfig<TFileContentType> {
  addGeneratedResources?: boolean;
  initOptions?: InitOptions;
  middleware?: I18nIndexMiddlewareConfig[];
  topOfFileComment?: string;
}

export interface I18nPluginDefaultNamespaceFileConfig
  extends Omit<I18nPluginFileGeneratorConfig, 'translationPathOrGetter' | 'translationWriter' | 'language'> {
  languages: string[];
  translationPathOrGetter?: I18nPluginDefinedAnySchemaTranslationPathGetter;
  translationWriter?: I18nPluginDefinedAnySchemaTranslationWriter;
}

export interface I18nPluginConfig extends IPluginConfig<I18nPluginFile> {
  conflictHandler: I18nPluginConflictHandler;
  // defaultNamespaceFile is where schema names for defined any values will go
  defaultNamespaceFile?: I18nPluginDefaultNamespaceFileConfig;
  files: I18nPluginFileGeneratorConfig[] | I18nPluginFileConfigCreator;
  indexFile?: I18nIndexFileConfig;
  namespaceWriter: NamespaceWriter;
}

export type I18nPluginConfigInput = Optional<I18nPluginConfig, 'conflictHandler' | 'namespaceWriter'>;

export class I18nPlugin extends BasePlugin<string, I18nPluginFileGeneratorConfig, I18nPluginFile, I18nPluginConfig, I18nPluginState> {
  name = 'I18nPlugin';
  private writtenTranslations: Record<string, Translation> = {};

  private static buildConfig(config: I18nPluginConfigInput): I18nPluginConfig {
    return {
      ...config,
      conflictHandler: config.conflictHandler ?? defaultConflictHandler,
      namespaceWriter: config.namespaceWriter ?? defaultNamespaceWriter,
    };
  }

  constructor(config: I18nPluginConfigInput) {
    super(I18nPlugin.buildConfig(config));
  }

  protected createPluginFilesFromConfig(fileConfig?: I18nPluginFileGeneratorConfig[]) {
    super.createPluginFilesFromConfig(
      (fileConfig || []).map((config) => ({ ...config, exportFromIndexFile: false, directory: `${config.directory}/${config.language}` })),
    );
  }

  private mergeAndBuildTranslations(newTranslations: Map<string, Translation>, existingTranslationsInFile: Map<string, Translation>) {
    const prospects = buildProspectiveTranslations(newTranslations, existingTranslationsInFile);

    const finalTranslationsForFile = new Map<string, Translation>();

    for (const [key, value] of prospects) {
      if (value.newValue !== undefined && value.existingValue !== undefined && value.newValue === value.existingValue) {
        finalTranslationsForFile.set(key, { key, value: value.newValue, source: value.source });
      } else if (value.newValue !== undefined || value.existingValue !== undefined) {
        const finalValue = this.pluginConfig.conflictHandler(value, prospects);

        if (finalValue) {
          finalTranslationsForFile.set(finalValue.key, finalValue);
        }
      }
    }

    const fileContent: ResourceLanguage = {};

    const translationsSortedByKeyName = sortByKey(Array.from(finalTranslationsForFile.values()), (entry) => entry.key);

    for (const translation of translationsSortedByKeyName) {
      setWith(fileContent, translation.key, translation.value, Object);
      this.writtenTranslations[translation.key] = translation;
    }

    return JSON.stringify(fileContent, null, 2);
  }

  public async run(): Promise<IPluginRunOutput<I18nPluginFile>> {
    for (const file of this.files) {
      let fileData: ResourceLanguage | undefined;

      try {
        fileData = parseExistingValue((await file.pollForExistingFileContent())?.content);
      } catch {}

      const existingTranslationsInFile = gatherTranslations(fileData);
      const newTranslations: Map<string, Translation> = new Map();

      for (const [, schema] of this.generatedSchemas) {
        if (file.isFileForSchema(schema)) {
          const translationPath =
            typeof file.config.translationPathOrGetter === 'function'
              ? file.config.translationPathOrGetter(schema)
              : file.config.translationPathOrGetter || defaultTranslationPathOrGetter(schema);

          if (translationPath) {
            const translationsForSchema = file.config.translationWriter
              ? file.config.translationWriter(schema, translationPath)
              : defaultSchemaTranslationWriter(schema, translationPath);

            for (const translation of translationsForSchema || []) {
              translation.source = schema;
              newTranslations.set(translation.key, translation);
            }
          }
        }
      }

      file.setRawContent(this.mergeAndBuildTranslations(newTranslations, existingTranslationsInFile));
    }

    if (this.pluginConfig.defaultNamespaceFile) {
      await this.generateDefaultNamespaceFiles();
    }

    if (this.pluginConfig.indexFile) {
      this.generateIndexFile();
    }

    const out = await this.buildFiles();

    return {
      files: out.reduce<IWritableFile[]>((acc, curr) => (curr ? [...acc, curr] : acc), []),
    };
  }

  private async generateDefaultNamespaceFile(language: string) {
    if (!this.pluginConfig.defaultNamespaceFile) {
      return;
    }

    const defaultNamespaceFile = this.createPluginFile(
      {
        ...this.pluginConfig.defaultNamespaceFile,
        directory: `${this.pluginConfig.defaultNamespaceFile.directory}/${language}`,
        language,
        exportFromIndexFile: false,
      },
      defaultGeneratorFileReader,
    );

    let fileData: ResourceLanguage | undefined;

    try {
      fileData = parseExistingValue((await defaultNamespaceFile.pollForExistingFileContent())?.content);
    } catch {}

    const existingTranslationsInFile = gatherTranslations(fileData);
    const newTranslations: Map<string, Translation> = new Map();

    for (const schemaName of this.definedAnySchemas) {
      const translationPath = (this.pluginConfig.defaultNamespaceFile.translationPathOrGetter || defaultDefinedAnySchemaTranslationPathGetter)(
        language,
        schemaName,
      );

      if (translationPath) {
        const translation = (this.pluginConfig.defaultNamespaceFile.translationWriter || defaultI18nPluginDefinedAnySchemaTranslationWriter)(
          language,
          schemaName,
          this.generatedSchemas.get(schemaName),
          translationPath,
        );

        if (translation) {
          newTranslations.set(translation.key, translation);
        }
      }
    }

    defaultNamespaceFile.setRawContent(this.mergeAndBuildTranslations(newTranslations, existingTranslationsInFile));
    this.files.push(defaultNamespaceFile);
  }

  private async generateDefaultNamespaceFiles() {
    if (!this.pluginConfig.defaultNamespaceFile || !this.pluginConfig.defaultNamespaceFile.languages.length) {
      return;
    }

    await Promise.all(this.pluginConfig.defaultNamespaceFile.languages.map((language) => this.generateDefaultNamespaceFile(language)));
  }

  private generateIndexFile() {
    if (!this.pluginConfig.indexFile) {
      return;
    }

    const { addGeneratedResources, initOptions, middleware, ...defaultFileConfig } = this.pluginConfig.indexFile;

    const indexFile = this.createPluginFile(
      {
        ...defaultFileConfig,
        exportFromIndexFile: false,
      },
      defaultGeneratorFileReader,
    );

    indexFile.addManualImport(I18NEXT_IMPORT_PATH, [], [], I18NEXT_DEFAULT_EXPORT_NAME);

    const resourcesByLanguageAndNamespace: Record<string, Record<string, string>> = {};
    const { resources = {}, ...remainingInitOptions } = initOptions || {};

    if (addGeneratedResources) {
      const filesToAdd = this.files.filter((file) => file.getHasContent());

      if (filesToAdd.length) {
        for (const file of filesToAdd) {
          if (!resourcesByLanguageAndNamespace[file.config.language]) {
            resourcesByLanguageAndNamespace[file.config.language] = {};
          }

          const languageResources = resourcesByLanguageAndNamespace[file.config.language]!;
          const namespaceName = file.config.namespaceName || this.pluginConfig.namespaceWriter(file);
          const resourceSpecifier = camelCase(`${file.config.language}_${namespaceName}_NS`);

          languageResources[namespaceName] = resourceSpecifier;

          indexFile.addImportToOtherGeneratedFile(file, undefined, undefined, resourceSpecifier);
        }
      }
    }

    const resourcesObjectLiteral = buildResourcesObjectLiteral(resources, resourcesByLanguageAndNamespace);

    // The index file is slightly different from the other files generated by this plugin, so just casting to any for now
    this.files.push(indexFile as any);

    let callExpression: ts.CallExpression | undefined;

    // Configure middleware
    for (const m of middleware || []) {
      indexFile.addManualImport(m.importPath, m.isDefault ? undefined : [m.importSpecifier], [], m.isDefault ? m.importSpecifier : undefined);

      callExpression = factory.createCallExpression(
        factory.createPropertyAccessExpression(
          callExpression || factory.createIdentifier(I18NEXT_DEFAULT_EXPORT_NAME),
          factory.createIdentifier(I18NEXT_USE_FUNCTION_NAME),
        ),
        undefined,
        [factory.createIdentifier(m.importSpecifier)],
      );
    }

    const builtLanguages = Array.from(new Set(Object.keys(resourcesByLanguageAndNamespace)));

    const resourcesAccess = factory.createIndexedAccessTypeNode(
      factory.createTypeQueryNode(factory.createIdentifier(I18N_INIT_OPTS_VAR_NAME)),
      factory.createLiteralTypeNode(factory.createStringLiteral('resources', true)),
    );

    if (this.pluginConfig.defaultNamespaceFile) {
      indexFile.addNodes(
        factory.createVariableStatement(
          [factory.createModifier(ts.SyntaxKind.ExportKeyword)],
          factory.createVariableDeclarationList(
            [
              factory.createVariableDeclaration(
                I18N_INIT_OPTS_VAR_NAME,
                undefined,
                undefined,
                factory.createAsExpression(
                  createObjectLiteral({
                    ...remainingInitOptions,
                    resources: resourcesObjectLiteral,
                    defaultNS: this.pluginConfig.defaultNamespaceFile.namespaceName,
                  }),
                  factory.createTypeReferenceNode('const'),
                ),
              ),
            ],
            ts.NodeFlags.Const,
          ),
        ),
        factory.createIdentifier('\n'),
      );
    }

    indexFile.addNodes(
      factory.createTypeAliasDeclaration(
        [factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        I18N_NAMESPACES_TYPE_NAME,
        undefined,
        factory.createUnionTypeNode(
          builtLanguages.map((lang) =>
            ts.factory.createTypeOperatorNode(
              ts.SyntaxKind.KeyOfKeyword,
              factory.createIndexedAccessTypeNode(resourcesAccess, factory.createLiteralTypeNode(factory.createStringLiteral(lang, true))),
            ),
          ),
        ),
      ),
      factory.createIdentifier('\n'),
    );

    callExpression = factory.createCallExpression(
      factory.createPropertyAccessExpression(
        callExpression || factory.createIdentifier(I18NEXT_DEFAULT_EXPORT_NAME),
        factory.createIdentifier(I18NEXT_INIT_FUNCTION_NAME),
      ),
      undefined,
      [factory.createIdentifier(I18N_INIT_OPTS_VAR_NAME)],
    );

    indexFile.addNodes(callExpression, factory.createIdentifier('\n'));

    indexFile.addManualExport(undefined, { namedExports: [I18NEXT_DEFAULT_EXPORT_NAME], typeOnlyExports: [] });
    indexFile.generateHeading();
  }

  getState(): I18nPluginState | undefined {
    return buildState(this.writtenTranslations);
  }
}
