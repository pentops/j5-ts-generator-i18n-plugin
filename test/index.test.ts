import { describe, it, expect } from 'vitest';
import { match, P } from 'ts-pattern';
import { camelCase, capitalCase, kebabCase, pascalCase } from 'change-case';
import {
  parseApiSource,
  GeneratedSchema,
  PackageSummary,
  APISource,
  defaultConfig,
  ParsedMethod,
  Builder,
  mergeConfig,
} from '@pentops/jsonapi-jdef-ts-generator';
import { I18nPlugin, I18nPluginFileGeneratorConfig } from '../src/plugin';
import { I18nPluginTranslationWriter, Translation } from '../src';
import mockApiSource from './helpers/mock-api.json';

function typeNameWriter(x: string) {
  return x
    .split(/[./]/)
    .filter((x) => x)
    .map((x) => pascalCase(x))
    .join('');
}

function methodNameWriter(method: ParsedMethod) {
  return method.fullGrpcName
    .split(/[./]/)
    .reduce<string[]>((acc, curr) => {
      if (curr) {
        acc.push(acc.length === 0 ? camelCase(curr) : pascalCase(curr));
      }

      return acc;
    }, [])
    .join('');
}

export const defaultMinorWords = [
  'and',
  'as',
  'but',
  'for',
  'if',
  'nor',
  'or',
  'so',
  'yet',
  'a',
  'an',
  'the',
  'as',
  'at',
  'by',
  'for',
  'in',
  'of',
  'off',
  'on',
  'per',
  'to',
  'up',
  'via',
  'with',
];

export function titleCaseName(name: string, forcedCaseWords: Record<string, string> = {}, minorWords: string[] = defaultMinorWords) {
  const capitalCased = capitalCase(name);
  const split = capitalCased.split(' ');

  return split
    .map((word, i) => {
      const lowerCasedWord = word.toLowerCase();

      if (forcedCaseWords[lowerCasedWord]) {
        return forcedCaseWords[lowerCasedWord];
      }

      if (i !== 0 && minorWords.includes(lowerCasedWord)) {
        return lowerCasedWord;
      }

      return word;
    })
    .join(' ');
}

function getPackageFileName(pkg: PackageSummary) {
  if (pkg.label) {
    return kebabCase(pkg.label.toLowerCase());
  }

  return kebabCase(pkg.package.split('.')[0].toLowerCase());
}

const caseOverrides: Record<string, string> = {
  api: 'API',
  id: 'ID',
  url: 'URL',
  jwt: 'JWT',
};

const i18nTranslationWriter: I18nPluginTranslationWriter = (
  schema: GeneratedSchema,
  schemaPath: string,
  existingValues?: Map<string, Translation>,
): Translation[] | undefined =>
  match(schema)
    .with({ rawSchema: { oneOf: P.not(P.nullish) } }, (s) =>
      Array.from(s.rawSchema.oneOf.properties.values()).map((property) => {
        const path = `${schemaPath}.${property.name}`;

        if (existingValues?.has(path)) {
          return existingValues.get(path)!;
        }

        return {
          key: path,
          value: titleCaseName(property.name, caseOverrides),
        };
      }),
    )
    .with({ rawSchema: { enum: P.not(P.nullish) } }, (s) =>
      s.rawSchema.enum.options.map((value) => {
        const path = `${schemaPath}.${value.name}`;

        if (existingValues?.has(path)) {
          return existingValues.get(path)!;
        }

        return {
          key: path,
          value: titleCaseName(value.name, caseOverrides),
        };
      }),
    )
    .otherwise(() => undefined);

describe(I18nPlugin, () => {
  const source = parseApiSource(mockApiSource as unknown as APISource);

  it('should write translations according to specified parameters', async () => {
    const p = new I18nPlugin({
      indexFile: {
        directory: './generated',
        fileName: 'index.ts',
        middleware: [{ importSpecifier: 'initReactI18next', importPath: 'react-i18next' }],
        addGeneratedResources: true,
      },
      files: (generatedSchemas) => {
        const directory = '../translation/translations/en';
        const files: Map<string, I18nPluginFileGeneratorConfig> = new Map();

        function getValidSchemaPackage(schema: GeneratedSchema) {
          return match(schema)
            .with(
              P.union({ rawSchema: { enum: { derivedHelperType: undefined } } }, { rawSchema: { oneOf: P.not(P.nullish) } }),
              (s) => s.parentPackage,
            )
            .otherwise(() => undefined);
        }

        function getFilterFunction(fullGrpcPackageName: string) {
          return (s: GeneratedSchema) => getValidSchemaPackage(s)?.package === fullGrpcPackageName;
        }

        for (const [, generatedSchema] of generatedSchemas) {
          const pkg = getValidSchemaPackage(generatedSchema);

          if (pkg) {
            const jsonName = getPackageFileName(pkg);

            if (!files.has(jsonName)) {
              files.set(jsonName, {
                directory,
                language: 'en',
                fileName: `${jsonName}.json`,
                translationWriter: i18nTranslationWriter,
                schemaFilter: getFilterFunction(pkg.package),
              });
            }
          }
        }

        return Array.from(files.values());
      },
    });

    const gen = await new Builder(
      process.cwd(),
      mergeConfig({
        dryRun: { log: false },
        typeOutput: {
          directory: './types/generated',
          fileName: 'api.ts',
        },
        clientOutput: {
          directory: './api-client/generated/client-functions',
          fileName: 'index.ts',
        },
        types: {
          enumType: 'enum',
          nameWriter: typeNameWriter,
        },
        client: {
          methodNameWriter,
        },
        plugins: [p as any],
      }),
      source,
    ).build();

    expect(gen).toBeDefined();

    const files = [...p.files].sort((a, b) => a.config.fileName.localeCompare(b.config.fileName));

    for (const file of files) {
      expect((file as any)._builtFile.writtenContent).toMatchSnapshot();
    }
  });
});
