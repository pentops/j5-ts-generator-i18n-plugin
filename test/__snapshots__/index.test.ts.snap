// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`I18nPlugin > should write translations according to specified parameters 1`] = `
"{
  "oneOf": {
    "O5DanteV1DeadMessageEventType": {
      "notified": "Notified",
      "rejected": "Rejected",
      "replayed": "Replayed",
      "updated": "Updated"
    }
  }
}"
`;

exports[`I18nPlugin > should write translations according to specified parameters 2`] = `
"/**
 * @generated by @pentops/jsonapi-jdef-ts-generator (Plugin: I18nPlugin) - do not edit */

import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import enDanteNs from '../../translation/translations/en/dante.json';
import enJ5Ns from '../../translation/translations/en/j5.json';
import enO5Ns from '../../translation/translations/en/o5.json';

i18n.use(initReactI18next).init({
    resources: {
        en: {
            dante: enDanteNs,
            o5: enO5Ns,
            j5: enJ5Ns
        }
    }
})

export { i18n };
"
`;

exports[`I18nPlugin > should write translations according to specified parameters 3`] = `
"{
  "oneOf": {
    "J5AuthV1AuthenticationMethod": {
      "external": "External",
      "jwt": "JWT",
      "session": "Session"
    }
  }
}"
`;

exports[`I18nPlugin > should write translations according to specified parameters 4`] = `
"{
  "oneOf": {
    "O5MessagingV1TopicProblem": {
      "unhandledError": "Unhandled Error"
    }
  }
}"
`;
